app/models/lecture.rb

```diff
@@ -1,3 +1,23 @@
+# needs to be at the top of the file
+class LectureValidator < ActiveModel::Validator
+  # changing attributes from a lecture is prohibited if the lecture ended
+  def validate(lecture)
+    if lecture.ended? && lecture.id
+      db_lecture = Lecture.find(lecture.id)
+      attributes_changed = lecture != db_lecture
+      # Additional id comparison is needed since you get the first entry in the db if the lecture has no id
+      if attributes_changed && lecture.id == db_lecture.id
+        (lecture_set_from_running_to_ended = lecture.ended?) && db_lecture.running?
+        other_attributes_than_status_changed = !lecture.compareIgnoreStatus(db_lecture)
+        # allow changing the lecture from running to ended
+        if !lecture_set_from_running_to_ended || other_attributes_than_status_changed
+          lecture.errors[:base] << "You cannot edit a lecture when is has been archived."
+        end
+      end
+    end
+  end
+end
```

Comment:
You introduce much code in this file, I think we could do this a little bit more concise.
E.g. we could not introduce a new validator class and the instance methods, but instead use the [readonly](https://apidock.com/rails/ActiveRecord/Base/readonly%3F) method.
So we could define
``` 
def readonly?
    if self.id
      db_lecture = Lecture.find(self.id)
      return true if db_lecture.status == "ended"
    end
    return false
end
```
to set the model instance to readonly. This makes the save call of the instance fail. We could test this with a small change to your test, see below.
What do you think?

Reply:
Nice. I did not find / know that. I tested if the `readonly`-method results preserves the same behavior and indeed it does. Since I agree that this looks way better I integrated your suggestion.
Thanks :smile: 
I just think that one of the lines can be written easier. Please check if you agree.
``` 
def readonly?
    if self.id
      db_lecture = Lecture.find(self.id)
      return db_lecture.status == "ended"
    end
    return false
end
```

Reply:
Great! Yes, you are right :)

Reply:
Please note that you have to revoke the changes that you requested. Otherwise (at least I think) the PR cannot be merged.

Reply:
Yea, and I just noticed, I  can't cite other lines that were changed before, after these review... :D 
Anyway, I think we don't need these functions anymore that were introduced like compareIgnoreStatus etc..  Or do you want to keep them for another reason?